commit 871dd2fa0a95c765390bb0602babf0f689e61fdb
Author: Jevin Sweval <jevinsweval@gmail.com>
Date:   Mon Dec 30 19:40:57 2024 -0500

    Add jeviterm 0.1.8 support

diff --git a/chardev/char-pty.c b/chardev/char-pty.c
index cbb21b76ae..ae361e72fb 100644
--- a/chardev/char-pty.c
+++ b/chardev/char-pty.c
@@ -35,6 +35,10 @@
 #include "chardev/char-io.h"
 #include "qom/object.h"
 
+#ifdef CONFIG_JEVITERM
+#include <jeviterm.h>
+#endif
+
 struct PtyChardev {
     Chardev parent;
     QIOChannel *ioc;
@@ -374,6 +378,19 @@ static void char_pty_open(Chardev *chr,
             s->path = g_strdup(path);
         }
     }
+
+#ifdef CONFIG_JEVITERM
+    if (backend->u.pty.data->spawngui) {
+        static int last_win_id = JEVITERM_NONE_WINDOW_ID;
+        const char *cmd = g_strdup_printf("/usr/bin/env bash -e -o pipefail -l -c \"/opt/homebrew/opt/picocom/bin/picocom -q %s; read\"", pty_name);
+        g_assert(cmd);
+        const char *cmds[] = {cmd, NULL};
+        const int new_win_id = jeviterm_open_tabs(cmds, 1, last_win_id, "qemu");
+        last_win_id = new_win_id;
+        g_free((void *)cmd);
+    }
+#endif
+
 }
 
 static void char_pty_parse(QemuOpts *opts, ChardevBackend *backend,
diff --git a/chardev/char.c b/chardev/char.c
index 44ff116fcd..c0a87aee5c 100644
--- a/chardev/char.c
+++ b/chardev/char.c
@@ -517,6 +517,8 @@ void qemu_chr_parse_common(QemuOpts *opts, ChardevCommon *backend)
     backend->logfile = g_strdup(logfile);
     backend->has_logappend = true;
     backend->logappend = qemu_opt_get_bool(opts, "logappend", false);
+
+    backend->spawngui = qemu_opt_get_bool(opts, "spawngui", false);
 }
 
 static const ChardevClass *char_get_class(const char *driver, Error **errp)
@@ -967,6 +969,10 @@ QemuOptsList qemu_chardev_opts = {
             .name = "abstract",
             .type = QEMU_OPT_BOOL,
 #endif
+        },{
+            .name = "spawngui",
+            .type = QEMU_OPT_BOOL,
+            .def_value_str = "off",
         },
         { /* end of list */ }
     },
diff --git a/meson.build b/meson.build
index f4109cd3ca..b5bf585914 100644
--- a/meson.build
+++ b/meson.build
@@ -1198,6 +1198,13 @@ if vmnet.found() and not cc.has_header_symbol('vmnet/vmnet.h',
   endif
 endif
 
+jeviterm = not_found
+if get_option('jeviterm').allowed()
+  jeviterm = dependency('jeviterm', version: '>=0.1.8',
+                        required: get_option('jeviterm'),
+                        method: 'pkg-config')
+endif
+
 seccomp = not_found
 seccomp_has_sysrawrc = false
 if not get_option('seccomp').auto() or have_system or have_tools
@@ -2446,6 +2453,7 @@ config_host_data.set('CONFIG_DARWIN', host_os == 'darwin')
 config_host_data.set('CONFIG_FDT', fdt.found())
 config_host_data.set('CONFIG_FUZZ', get_option('fuzzing'))
 config_host_data.set('CONFIG_GCOV', get_option('b_coverage'))
+config_host_data.set('CONFIG_JEVITERM', jeviterm.found())
 config_host_data.set('CONFIG_LIBUDEV', libudev.found())
 config_host_data.set('CONFIG_LINUX', host_os == 'linux')
 config_host_data.set('CONFIG_POSIX', host_os != 'windows')
@@ -3988,11 +3996,11 @@ qmp = declare_dependency(objects: libqmp.extract_all_objects(recursive: false),
                          dependencies: qmp_ss.dependencies())
 
 libchardev = static_library('chardev', chardev_ss.sources() + genh,
-                            dependencies: chardev_ss.dependencies(),
+                            dependencies: chardev_ss.dependencies() + jeviterm,
                             build_by_default: false)
 
 chardev = declare_dependency(objects: libchardev.extract_all_objects(recursive: false),
-                             dependencies: chardev_ss.dependencies())
+                             dependencies: chardev_ss.dependencies() + jeviterm)
 
 hwcore_ss = hwcore_ss.apply({})
 libhwcore = static_library('hwcore', sources: hwcore_ss.sources() + genh,
@@ -4702,6 +4710,7 @@ summary(summary_info, bool_yn: true, section: 'Crypto')
 summary_info = {}
 if host_os == 'darwin'
   summary_info += {'Cocoa support':           cocoa}
+  summary_info += {'jeviterm support':        jeviterm}
 endif
 summary_info += {'SDL support':       sdl}
 summary_info += {'SDL image support': sdl_image}
diff --git a/meson_options.txt b/meson_options.txt
index 5eeaf3eee5..e92d412943 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -148,6 +148,8 @@ option('bpf', type : 'feature', value : 'auto',
         description: 'eBPF support')
 option('cocoa', type : 'feature', value : 'auto',
        description: 'Cocoa user interface (macOS only)')
+option('jeviterm', type : 'feature', value : 'auto',
+       description: 'jeviterm-based iTerm2 integration (macOS only)')
 option('curl', type : 'feature', value : 'auto',
        description: 'CURL block device driver')
 option('gio', type : 'feature', value : 'auto',
diff --git a/qapi/char.json b/qapi/char.json
index e045354350..811cc8631c 100644
--- a/qapi/char.json
+++ b/qapi/char.json
@@ -199,11 +199,14 @@
 # @logappend: true to append instead of truncate (default to false to
 #     truncate)
 #
+# @spawngui: true to spawn a GUI terminal attached to the chardev(s)
+#
 # Since: 2.6
 ##
 { 'struct': 'ChardevCommon',
   'data': { '*logfile': 'str',
-            '*logappend': 'bool' } }
+            '*logappend': 'bool',
+            '*spawngui': 'bool'} }
 
 ##
 # @ChardevFile:
diff --git a/qemu-options.hx b/qemu-options.hx
index cc694d3b89..9f50f047ff 100644
--- a/qemu-options.hx
+++ b/qemu-options.hx
@@ -3696,7 +3696,7 @@ DEF("chardev", HAS_ARG, QEMU_OPTION_chardev,
     "-chardev console,id=id[,mux=on|off][,logfile=PATH][,logappend=on|off]\n"
     "-chardev serial,id=id,path=path[,mux=on|off][,logfile=PATH][,logappend=on|off]\n"
 #else
-    "-chardev pty,id=id[,path=path][,mux=on|off][,logfile=PATH][,logappend=on|off]\n"
+    "-chardev pty,id=id[,path=path][,mux=on|off][,logfile=PATH][,logappend=on|off][,spawngui=on|off]\n"
     "-chardev stdio,id=id[,mux=on|off][,signal=on|off][,logfile=PATH][,logappend=on|off]\n"
 #endif
 #ifdef CONFIG_BRLAPI
diff --git a/scripts/meson-buildoptions.sh b/scripts/meson-buildoptions.sh
index a8066aab03..70ca72d008 100644
--- a/scripts/meson-buildoptions.sh
+++ b/scripts/meson-buildoptions.sh
@@ -133,6 +133,7 @@ meson_options_help() {
   printf "%s\n" '  hvf             HVF acceleration support'
   printf "%s\n" '  iconv           Font glyph conversion support'
   printf "%s\n" '  jack            JACK sound support'
+  printf "%s\n" '  jeviterm        jeviterm-based iTerm2 integration (macOS only)'
   printf "%s\n" '  keyring         Linux keyring support'
   printf "%s\n" '  kvm             KVM acceleration support'
   printf "%s\n" '  l2tpv3          l2tpv3 network backend support'
@@ -356,6 +357,8 @@ _meson_option_parse() {
     --interp-prefix=*) quote_sh "-Dinterp_prefix=$2" ;;
     --enable-jack) printf "%s" -Djack=enabled ;;
     --disable-jack) printf "%s" -Djack=disabled ;;
+    --enable-jeviterm) printf "%s" -Djeviterm=enabled ;;
+    --disable-jeviterm) printf "%s" -Djeviterm=disabled ;;
     --enable-keyring) printf "%s" -Dkeyring=enabled ;;
     --disable-keyring) printf "%s" -Dkeyring=disabled ;;
     --enable-kvm) printf "%s" -Dkvm=enabled ;;
